exercicios/para-casa/entregas/exemplo-nome-sobrenome/jeanne-mendes/projeto/Conta/Conta.js

```diff
@@ -0,0 +1,183 @@
+    }
+  }
+  sacar(valor) {
+    if (valor > 0 && typeof valor === "number") {
+      if (this.#saldo - valor > 0) {
+        const saldoAtualizado = this.#saldo - valor;
+        this.setSaldo(saldoAtualizado);
+      } else {
+        throw new Error("Saldo insuficiente");
+      }
+    } else {
+      throw new Error("Valor inválido para saque");
+    }
+  }
+
+  depositar(valor) {
+    if (valor > 0 && typeof valor === "number") {
+      const saldoAtualizado = this.#saldo + valor;
+      this.setSaldo(saldoAtualizado);
+    } else {
+      throw new Error("Valor inválido para depósito");
+    }
+  }
+
+  transferir(valor, agencia, conta) {
+    //LISTA.find(APELIDO PARA ITEM SELECIONADA => COMPARACAO )
+    /**
+     * antes de fazer a transferencia preciso verificar se a conta receptora existe na lista de contas
+     * contaValida vai me retornar a conta se ela existir e undefined se não existir
+     */
+    let contaValida = Conta.listaContas.find((contaReceptora) => {
+      let numeroContaReceptora = contaReceptora.getConta();
+      let numeroAgenciaReceptora = contaReceptora.getAgencia();
+      return (
+        numeroContaReceptora === conta && numeroAgenciaReceptora === agencia
+      );
+    });
+
+    if (!contaValida) {
+      throw new Error("Conta não encontrada");
+    }
+
+    if (valor < 0) {
+      throw new Error("Valor inválido para transferencia");
+    }
+
+    //a conta não pode ficar negativa ao fazer a transferencia,
+    if (this.#saldo - valor > 0) {
+      const saldoAtualizado = this.#saldo - valor;
+      this.setSaldo(saldoAtualizado);
+      const saldoContaReceptora = contaValida.getSaldo() + valor;
+      contaValida.setSaldo(saldoContaReceptora);
+      return "Tranferencia realizada";
+    }
+
+    if (valor > this.#saldo) {
+      return "Saldo insuficiente.";
+    }
+  }
+  pix(valor, chavePix, tipo) {
+    let contaValida = Conta.listaContas.find(
+      (conta) => conta.chavesPix[tipo] === chavePix
+    );
+
+    if (!contaValida) {
+      throw new Error("Chave pix não encontrada");
+    }
+
+    if (valor < 0) {
+      throw new Error("Valor inválido para transferencia");
+    }
+
+    //a conta não pode ficar negativa ao fazer a transferencia,
+    if (this.#saldo - valor > 0) {
+      const saldoAtualizado = this.#saldo - valor;
+      this.setSaldo(saldoAtualizado);
+      const saldoContaReceptora = contaValida.getSaldo() + valor;
+      contaValida.setSaldo(saldoContaReceptora);
+      return "Tranferencia realizada";
+    }
+
+    if (valor > this.#saldo) {
+      return "Saldo insuficiente.";
+    }
+  }
+
+  getAgencia() {
+    return this.#agencia;
+  }
+
+  getConta() {
+    return this.#conta;
+  }
+
+  getSaldo() {
+    return this.#saldo;
+  }
+
+  setAgencia(agencia) {
+    this.#agencia = agencia;
+  }
+
+  setConta(conta) {
+    this.#conta = conta;
+  }
+
+  setSaldo(novoSaldo) {
+    this.#saldo = novoSaldo;
+  }
+
+  criarChavePix(chavePix, tipo) {
```

Comment:
Seguindo o principio de legibilidade podemos melhorar esse código criando uma função auxiliar. Aprendi isso no meu trabalho, veja o que acha 😃

`criarChavePix(chavePix, tipo) {
  const regexMap = {
    CPF: /([0-9]{2}[\.]?[0-9]{3}[\.]?[0-9]{3}[\/]?[0-9]{4}[-]?[0-9]{2})|([0-9]{3}[\.]?[0-9]{3}[\.]?[0-9]{3}[-]?[0-9]{2})/,
    EMAIL: /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/,
    TELEFONE: /^(?:(?:\+|00)?(55)\s?)?(?:\(?([1-9][0-9])\)?\s?)?(?:((?:9\d|[2-9])\d{3})\-?(\d{4}))$/,
  };

  if (!regexMap[tipo]) {
    throw new Error("Tipo de chavePix inválido");
  }

  if (regexMap[tipo].test(chavePix)) {
    this.chavesPix[tipo.toLowerCase()] = chavePix;
    return `Chave Pix por ${tipo.toLowerCase()} criada com sucesso`;
  } else {
    throw new Error(`Erro: ${tipo} inválido`);
  }
}`
o código fica mais conciso, modular e fácil de entender.

Reply:
`criarChavePix(chavePix, tipo) {
  const regexMap = {
    CPF: /([0-9]{2}[\.]?[0-9]{3}[\.]?[0-9]{3}[\/]?[0-9]{4}[-]?[0-9]{2})|([0-9]{3}[\.]?[0-9]{3}[\.]?[0-9]{3}[-]?[0-9]{2})/,
    EMAIL: /^\w+([\.-]?\w+)*@\w+([\.-]?\w+)*(\.\w{2,3})+$/,
    TELEFONE: /^(?:(?:\+|00)?(55)\s?)?(?:\(?([1-9][0-9])\)?\s?)?(?:((?:9\d|[2-9])\d{3})\-?(\d{4}))$/,
  };

  if (!regexMap[tipo]) {
    throw new Error("Tipo de chavePix inválido");
  }

  if (regexMap[tipo].test(chavePix)) {
    this.chavesPix[tipo.toLowerCase()] = chavePix;
    return `Chave Pix por ${tipo.toLowerCase()} criada com sucesso`;
  } else {
    throw new Error(`Erro: ${tipo} inválido`);
  }
}
`

Reply:
Nossa que top, já estou fatorando seguindo todas suas dicas, muito obrigada ❤️